tools/imatrix/imatrix.cpp

```diff
@@ -125,95 +128,241 @@ static void process_tensor_name(const std::string & input, std::string & layer,
+            for (int j = 0; j < row_size; ++j) {
+                activations.push_back(e.values[i*row_size + j] / e.counts[i]);
+            }
+        }
+    } else {
+        activations.reserve(e.activations.size());
+
+        for (int i = 0; i < n_mat; ++i) {
+            for (int j = 0; j < row_size; ++j) {
+                activations.push_back(e.activations[i*row_size + j] / e.counts[i]);
+            }
         }
     }
 
-    const float act_total     = std::accumulate(activations.begin(), activations.end(), 0.0f);
-    const float act_max       = *std::max_element(activations.begin(), activations.end());
-    const float act_min       = *std::min_element(activations.begin(), activations.end());
-    const float act_mean      = act_total / activations.size();
-    const float act_sqr_total = std::inner_product(activations.begin(), activations.end(), activations.begin(), 0.0f);
-    const float act_var       = (act_sqr_total / activations.size()) - (act_mean * act_mean);
-    const float act_dev       = std::sqrt(std::max(0.0f, act_var));
-    float threshold           = 1e-5f;
-    const int inactive_count  = std::count_if(activations.begin(), activations.end(),
-                                               [threshold](const float v) { return fabsf(v) <= threshold; });
-    const float active_ratio  = 1 - static_cast<float>(inactive_count) / activations.size();
+    const float sum             = std::accumulate(activations.begin(), activations.end(), 0.0f);
+    const float max             = *std::max_element(activations.begin(), activations.end());
+    const float min             = *std::min_element(activations.begin(), activations.end());
+    const float mean            = sum / activations.size();
+    const float sqr_sum         = std::inner_product(activations.begin(), activations.end(), activations.begin(), 0.0f);
+    const float variance        = (sqr_sum / activations.size()) - (mean * mean);
+    const float std_deviation   = std::sqrt(std::max(0.0f, variance));
+    float entropy               = 0;
+
+    if (e.activations.empty()) {
+        if (sum > 0) {
+            for (const auto act : activations) {
+                if (const float p = act / sum; p > 0) {
+                    entropy -= p * std::log2(p);
+                }
+            }
+        }
+    } else {
+        float div = 0.0;
+        std::vector<float> weights(activations.size());
+        for (size_t i = 0; i < activations.size(); ++i) {
+            const float w = activations[i] * activations[i];
+            weights[i] = w;
+            div += w;
+        }
 
-    float entropy = 0;
-    if (act_total > 0) {
-        for (const auto act : activations) {
-            if (const float p = act / act_total; p > 0) {
-                entropy -= p * std::log2(p);
+        if (div > 0.0) {
+            for (float w : weights) {
+                const float p = w / div;
+                if (p > 0.0) entropy -= p * std::log2(p);
```

Comment:
(style)

```suggestion
                if (p > 0.0) { entropy -= p * std::log2(p); }
```

(You can also split it on multiple lines, depending on your preference)